src/halfinteger.jl

```diff
@@ -1,45 +1,77 @@
 # HalfInteger
+"""
+    struct HalfInteger <: Real
+
+Represents half-integer values.
+"""
 struct HalfInteger <: Real
-    num::Int
+    twofold::Int
+
+    """
+        HalfInteger(numerator::Integer, denominator::Integer)
+
+    Constructs a `HalfInteger` object as a rational number from the given integer numerator
+    and denominator values.
+    """
+    function HalfInteger(num::Integer, den::Integer)
+        (den == 2) && return new(num)
+        (den == 1) && return new(2*num)
+        (den == 0) && throw(ArgumentError("Denominator can not be zero."))
+        # If non-trivial, we'll see if we can reduce it down to a half-integer
+        twofold, r = divrem(2*num, den)
+        if r == 0
+            return new(twofold)
+        else
+            throw(ArgumentError("$num // $den is not a half-integer value."))
+        end
+    end
 end
-Base.:+(a::HalfInteger, b::HalfInteger) = HalfInteger(a.num+b.num)
-Base.:-(a::HalfInteger, b::HalfInteger) = HalfInteger(a.num-b.num)
-Base.:-(a::HalfInteger) = HalfInteger(-a.num)
-Base.:<=(a::HalfInteger, b::HalfInteger) = a.num <= b.num
-Base.:<(a::HalfInteger, b::HalfInteger) = a.num < b.num
-Base.one(::Type{HalfInteger}) = HalfInteger(2)
-Base.zero(::Type{HalfInteger}) = HalfInteger(0)
+HalfInteger(x::Real) = convert(HalfInteger, x)
 
 Base.promote_rule(::Type{HalfInteger}, ::Type{<:Integer}) = HalfInteger
 Base.promote_rule(::Type{HalfInteger}, T::Type{<:Rational}) = T
 Base.promote_rule(::Type{HalfInteger}, T::Type{<:Real}) = T
 
-Base.convert(::Type{HalfInteger}, n::Integer) = HalfInteger(2*n)
+Base.convert(::Type{HalfInteger}, n::Integer) = HalfInteger(2*n, 2)
 function Base.convert(::Type{HalfInteger}, r::Rational)
     if r.den == 1
-        return HalfInteger(2*r.num)
+        return HalfInteger(2*r.num, 2)
     elseif r.den == 2
-        return HalfInteger(r.num)
+        return HalfInteger(r.num, 2)
     else
         throw(InexactError(:HalfInteger, HalfInteger, r))
     end
 end
 function Base.convert(::Type{HalfInteger}, r::Real)
     num = 2*r
```

Comment:
Maybe also call it `twofold` then, although I did also like `num` for analogy with `Rational` :-)

Reply:
I can also switch the name of the field back to `num`. My reason for the change was that it took me ages to realize that `num` stands for "numerator".. I assumed "number" or something like that. I think it's because, unlike for the `Rational` type, there is no explicit `den` anywhere in sight.

`twofold` is definitely not ideal, but it was the best I was able to come up with at the time that sounded more descriptive for me. However, now that I realized that analogy with `Rational`, I also like that. I can think of three options:

* Rename back to `num`, but have a comment next to it.
* Spell it out as `numerator` instead.
* Keep it as `twofold` (and update the variable you pointed out here).

Reply:
Anyway you like, I have no strong preference.

Reply:
I went with calling it `numerator`.